frame/contracts/src/gas.rs

```diff
@@ -181,17 +185,44 @@ impl<T: Config> GasMeter<T> {
 		self.gas_left = self.gas_left.saturating_add(adjustment).min(self.gas_limit);
 	}
 
+	/// This method is used for gas syncs with the engine.
+	///
+	/// Updates internal `engine_comsumed` tracker of engine fuel consumption.
+	///
+	/// Charges self with the `ref_time` Weight corresponding to wasmi fuel consumed on the engine
+	/// side since last sync. Passed value is scaled by multiplying it by the weight of a basic
+	/// operation, as such an operation in wasmi engine costs 1.
+	///
+	/// Returns the updated `gas_left` `Weight` value from the meter.
+	/// Normally this would never fail, as engine should fail first when out of gas.
+	pub fn sync_fuel(&mut self, wasmi_fuel: u64) -> Result<Weight, DispatchError> {
+		let wasmi_fuel = wasmi_fuel.saturating_sub(self.engine_consumed);
```

Comment:
> Why do we need to subtract?

Because there could be several host function calls in a single contract call. See this [test](https://github.com/paritytech/substrate/blob/ag-wasmeter/frame/contracts/src/tests.rs#L910) for an example.